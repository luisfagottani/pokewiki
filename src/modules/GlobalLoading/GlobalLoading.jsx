import classNames from 'classnames';

import { ReactComponent as PokeballIcon } from 'assets/images/pokeball-icon.svg';

import style from './GlobalLoading.module.scss';

const GlobalLoading = () => (
  <div className={classNames(style.loading, style['loading--hide'])}>
    <div className={style.loading__stage}></div>
    <div className={classNames(style.loading__stage, style['loading__stage--bottom'])}></div>
    <PokeballIcon className={style.loading__ball} />
  </div>
);

export default GlobalLoading;
