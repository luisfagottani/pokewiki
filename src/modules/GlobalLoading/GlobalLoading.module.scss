.loading {
  // Position
  position: fixed;
  top: 0;
  left: 0;
  z-index: 10;

  // Box Model
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;

  // Visual
  animation: show 300ms backwards;

  &--hide {
    // Visual
    animation: hide 1000ms forwards;
  }

  &__ball {
    // Box Model
    width: 100px;
    height: 100px;
    margin: auto;

    // Position
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;

    // Visual
    animation: tada 1000ms infinite;

    .loading--hide & {
      // Visual
      animation: hideBall 0.8s ease-in-out forwards;
    }
  }

  &__stage {
    // Box Model
    width: 100%;
    height: 50%;

    // Visual
    background-color: #fff;

    .loading--hide & {
      // Visual
      animation: stageUp 1000ms forwards;
    }

    &--bottom {
      // Visual
      background: linear-gradient(to bottom, #d10000 0%, #ff0000 50%);

      .loading--hide & {
        // Visual
        animation: stageDown 1000ms forwards;
      }
    }
  }
}

@keyframes tada {
  0% {
    transform: scale(1);
  }
  10%,
  20% {
    transform: scale(0.9) rotate(-3deg);
  }
  30%,
  50%,
  70%,
  90% {
    transform: scale(1.1) rotate(3deg);
  }
  40%,
  60%,
  80% {
    transform: scale(1.1) rotate(-3deg);
  }
  100% {
    transform: scale(1) rotate(0);
  }
}

@keyframes stageDown {
  0% {
    transform: translateY(0);
  }

  100% {
    transform: translateY(100%);
  }
}

@keyframes stageUp {
  0% {
    transform: translateY(0);
  }

  100% {
    transform: translateY(-100%);
  }
}

@keyframes hideBall {
  0% {
    transform: scale(1) rotate(0deg);
  }

  100% {
    transform: scale(0) rotate(360deg);
  }
}

@keyframes hide {
  0% {
    z-index: 10;
    opacity: 1;
  }

  100% {
    z-index: -1;
    opacity: 0;
  }
}

@keyframes show {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
    z-index: 10;
  }
}
