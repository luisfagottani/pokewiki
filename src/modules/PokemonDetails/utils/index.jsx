const progressBarColor = (progress) => {
  if (progress < 60) {
    return 'blue';
  } else if (progress >= 60 && progress <= 70) {
    return 'yellow';
  } else {
    return 'red';
  }
};

export { progressBarColor };
