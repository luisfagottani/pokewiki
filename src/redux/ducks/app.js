import produce from 'immer';

export const REDUCER_NAME = 'app';

// TYPES
export const Types = {
  SET_GLOBAL_LOADING: `${REDUCER_NAME}/SET_GLOBAL_LOADING`,
};

// INITIAL STATE
const initialState = {
  isLoading: true,
  locale: 'pt-BR',
};

// REDUCER
export const reducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case Types.SET_GLOBAL_LOADING:
      draft.isLoading = action.payload;
      return draft;

    case Types.SET_USER_PROFILE:
      draft.profile = action.profile;
      return draft;

    default:
      return draft;
  }
});

// ACTION FUNCTIONS

export function defineGlobalLoading(status) {
  return {
    type: Types.SET_GLOBAL_LOADING,
    payload: status,
  };
}

// SELECTOR FUNCTIONS
export function getGlobalLoading(state) {
  return state[REDUCER_NAME].isLoading;
}

export function getLocale(state) {
  return state[REDUCER_NAME].locale;
}
