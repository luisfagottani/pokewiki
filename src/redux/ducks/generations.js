import produce from 'immer';

export const REDUCER_NAME = 'family';

// TYPES
export const Types = {
  FETCH_GENERATIONS: `${REDUCER_NAME}/FETCH_POKEMONS`,
  FETCH_GENERATIONS_BY_ID: `${REDUCER_NAME}/FETCH_GENERATIONS_BY_ID`,
  PERSIST_GENERATIONS: `${REDUCER_NAME}/PERSIST_GENERATIONS`,
};

// INITIAL STATE
const initialState = {
  generations: [],
};

// REDUCER
export const reducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case Types.PERSIST_GENERATIONS:
      draft.generations = { ...action.payload };
      return draft;

    default:
      return draft;
  }
});

// ACTION FUNCTIONS
export function fetchGenerations() {
  return {
    type: Types.FETCH_GENERATIONS,
  };
}

export function fetchGenerationsById(id) {
  return {
    type: Types.FETCH_GENERATIONS_BY_ID,
    payload: id,
  };
}

export function persistGenerations(pokemons) {
  return {
    type: Types.PERSIST_GENERATIONS,
    payload: pokemons,
  };
}

// SELECTORS
export function getGenerations(state) {
  return state[REDUCER_NAME]?.generations;
}
