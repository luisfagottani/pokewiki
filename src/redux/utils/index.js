const reduceSpecies = (results) =>
  results.reduce((current, next) => {
    const name = next?.name;
    if (!current[name]) {
      current[name] = { name };
    }
    return current;
  }, {});

export { reduceSpecies };
