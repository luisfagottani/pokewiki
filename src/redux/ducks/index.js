import { combineReducers } from 'redux';

import { reducer as AppReducer, REDUCER_NAME as APP_REDUCER_NAME } from './app';
import { reducer as PokemonsReducer, REDUCER_NAME as APP_POKEMONS_REDUCER } from './pokemons';
import { reducer as GenerationsReducer, REDUCER_NAME as APP_GENERATIONS_REDUCER } from './generations';

export default combineReducers({
  [APP_REDUCER_NAME]: AppReducer,
  [APP_POKEMONS_REDUCER]: PokemonsReducer,
  [APP_GENERATIONS_REDUCER]: GenerationsReducer,
});
