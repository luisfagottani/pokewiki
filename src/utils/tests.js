import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import { createStore, combineReducers } from 'redux';
import i18nMessages from 'i18n';

import middleware, { sagaMiddleware } from 'redux/middleware';

export const renderWithReactIntl = (component) => (
  <IntlProvider messages={i18nMessages['pt-BR']} locale="pt-BR">
    {component}
  </IntlProvider>
);

export const componentWithRouter = (initialRoute) => (children) => {
  if (typeof children === 'function') {
    return (
      <MemoryRouter initialIndex={0} initialEntries={initialRoute && [initialRoute]}>
        {children()}
      </MemoryRouter>
    );
  }
  return (
    <MemoryRouter initialEntries={initialRoute && [initialRoute]} initialIndex={0}>
      {children}
    </MemoryRouter>
  );
};

export const createMockStore = ({ reducers, initialState }) => {
  return createStore(combineReducers(reducers), initialState, middleware);
};

export const mockSaga = (fnGenerators) => sagaMiddleware.run(fnGenerators);

function render(ui, store, { ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
