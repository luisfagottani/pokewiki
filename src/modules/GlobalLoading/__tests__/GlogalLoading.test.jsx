import React from 'react';
import { createMockStore, render } from 'utils/tests';
import { defineGlobalLoading, reducer as AppReducer, REDUCER_NAME as APP_REDUCER_NAME } from 'redux/ducks/app';
import GlobalLoading from '..';

const mockStore = createMockStore({
  reducers: { [APP_REDUCER_NAME]: AppReducer },
  initialState: {
    [APP_REDUCER_NAME]: {
      isLoading: true,
    },
  },
});

describe('<GlobalLoading />', () => {
  it('Should render without errors', () => {
    const { container } = render(<GlobalLoading />, mockStore);
    expect(container).toMatchSnapshot();
  });

  describe('When the loading state is equal to true', () => {
    it('should render without the hide class', () => {
      const { container } = render(<GlobalLoading />, mockStore);

      expect(container.firstChild).not.toHaveClass('loading--hide');
    });
  });

  describe('When the loading state is equal to false', () => {
    it('should render with the hide class', () => {
      const { container, store } = render(<GlobalLoading />, mockStore);

      store.dispatch(defineGlobalLoading(false));

      expect(container.firstChild).toHaveClass('loading--hide');
    });
  });
});
