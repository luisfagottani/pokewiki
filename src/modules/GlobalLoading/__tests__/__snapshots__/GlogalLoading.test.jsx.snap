// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalLoading /> Should render without errors 1`] = `
<div>
  <div
    class="loading"
  >
    <div
      class="loading__stage"
    />
    <div
      class="loading__stage loading__stage--bottom"
    />
    <svg
      class="loading__ball"
    >
      pokeball-icon.svg
    </svg>
  </div>
</div>
`;
